src/gc/heap.cpp

```diff
@@ -137,6 +139,68 @@ void registerGCManagedBytes(size_t bytes) {
     }
 }
 
+//////
+/// Finalizers
+
+static const std::string _del_str("__del__");
+
+bool hasFinalizer(Box* b) {
+    if (b->cls == instance_cls) {
+        // For old-style classes, pretend that there is always a finalizer. We do this
+        // because old-style classes don't use tp_del, so looking for the finalizer is
+        // always an attribute lookup.
+        // TODO: Check if doing the attribute lookup would give performance gains.
+        return true;
+    } else {
+        return b->cls->tp_del || b->cls->hasNativeDestructor();
+    }
+}
+
+bool hasOrderedFinalizer(Box* b) {
+    if (b->cls == instance_cls) {
+        return true;
+    } else {
+        return b->cls->tp_del || (!b->cls->has_safe_tp_dealloc && b->cls->hasNativeDestructor());
+    }
+}
+
+void finalize(Box* b) {
+    GCAllocation* al = GCAllocation::fromUserData(b);
+    assert(!hasFinalized(al));
+    setFinalized(al);
+    b->cls->tp_dealloc(b);
+}
+
+
+void finalizeIfUnordered(Box* b) {
+    if (!hasOrderedFinalizer(b)) {
+        GCAllocation* al = GCAllocation::fromUserData(b);
+        if (!hasFinalized(al)) {
+            finalize(b);
+        }
+    }
+}
+
+bool isWeakReference(Box* b) {
```

Comment:
this should probably be called `isWeaklyReferenced`.  `isWeakReference` sounds like we're testing the box to see if it's an instance of a WeakRef type.